src/oc_erchef/apps/chef_index/src/chef_wait_group.erl

```diff
@@ -0,0 +1,205 @@
+                                                         waiting_workers = _StillWaiting}) ->
+    {noreply, State#chef_wait_group_state{waiting_pid = From}};
+%% If someone is already waiting, error
+handle_call(gather, _From, State = #chef_wait_group_state{waiting_pid = _Pid}) ->
+    {reply, {error, already_waiting}, State};
+%% If a job is done, remove it from the waiting list.  If this is the
+%% last job, rely to the waiting pid if we have one.
+handle_call({job_done, Result}, {Pid, _Tag}, State) ->
+    State1 = mark_job_done(Pid, Result, State),
+    case maybe_reply_to_waiter(State1) of
+        true ->
+            {stop, normal, ok, #chef_wait_group_state{}};
+        false ->
+            {reply, ok, State1}
+    end;
+handle_call({job_failed, Result}, {Pid, _Tag}, State) ->
+    State1 = mark_job_failed(Pid, Result, State),
+    case maybe_reply_to_waiter(State1) of
+        true ->
+            {stop, normal, ok, #chef_wait_group_state{}};
+        false ->
+            {reply, ok, State1}
+    end;
+handle_call(get_state, _From, State) ->
+    {reply, State, State}.
+
+handle_info({'EXIT', _From, normal}, State) ->
+    {noreply, State};
+handle_info({'EXIT', From, Reason}, State) ->
+    lager:error("Worker ~p failed unexpectedly: ~p", [From, Reason]),
+    State1 = mark_job_failed(From, {worker_failed, Reason}, State),
+    case maybe_reply_to_waiter(State1) of
+        true ->
+            {stop, normal, State1};
+        false ->
+            {noreply, State1}
+    end.
+
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+spawn_worker(JobName, JobArgs, State = #chef_wait_group_state{job_fun = Fun,
+                                                              waiting_workers = Workers}) ->
+    Parent = self(),
+    Pid = spawn_link(fun() -> run_user_callback(Fun, JobArgs, Parent) end),
+    {ok, State#chef_wait_group_state{waiting_workers = [{Pid, JobName}|Workers]}}.
+
+run_user_callback(Fun, JobArgs, Parent) ->
+    try apply(Fun, JobArgs) of
+        Res -> gen_server:call(Parent, {job_done, Res})
+    catch
+        Exception:Reason -> gen_server:call(Parent, {job_failed, {Exception, Reason}})
+    end.
+
+-spec maybe_reply_to_waiter(#chef_wait_group_state{}) -> boolean().
+maybe_reply_to_waiter(#chef_wait_group_state{waiting_pid = undefined}) ->
+    false;
+maybe_reply_to_waiter(#chef_wait_group_state{waiting_pid = Pid,
+                                             waiting_workers = [],
+                                             done_jobs = Done,
+                                             failed_jobs = Failed}) ->
+    gen_server:reply(Pid, make_gather_reply(Done, Failed)),
```

Comment:
Is it possible that `waiting_pid` has gone away? (I don't think it's expected in the current set up, but if this would be used more generically, would we have to deal with that here?)